mlflow/store/artifact/sftp_artifact_repo.py

```diff
@@ -28,12 +28,16 @@ def __init__(self, connections):
         self._cond = threading.Condition()
 
     @contextmanager
-    def get_sfp_connection(self):
+    def get_sftp_connection(self):
         with self._cond:
             self._cond.wait_for(lambda: bool(self._connections))
             connection = self._connections.pop(-1)
+
         yield connection
-        self._connections.append(connection)
+
+        with self._cond:
+            self._connections.append(connection)
+            self._cond.notify()
```

Comment:
If you're saturating the threadpool and only a single thread is available to receive a notification, if multiple threads are waiting on the notification, couldn't the deadlock still exist potentially? 

Would `self._cond.notify_all()` be a safer approach?

Reply:
I don't have a lot of experience on the Python threading module, how could a single thread be available to receive the notification and multiple threads are waiting the notification, simultaneously?
I think it's also safe to go for the `self._cond.notify_all()`, both should work, but when notifying all threads, they all would compete for the lock, which I don't think is necessary. For this use case, we can simply notify a single waiting thread and move on. We don't need to worry about fairness and such, we can safely assume a limited number of files exists and they all eventually get their connection.
I can test it with `self._cond.notify_all()`, to see if I get different results, or even stress test with the proposed `self._cond.notify()` to check for potential deadlock in the situation you mentioned, of a saturated thread pool.

Reply:
new PR: https://github.com/mlflow/mlflow/pull/12938